### PR TITLE
Implement user email change functionality

### LINKED ISSUE #1996: Allow Changing Emails
Currently, we have users email an admin email to change their email. We could, and really, should, automate this process, by sending a verification to both emails... Or maybe just the new one?

We should consider a case where a user has lost access to their old email but still has access to their account through, say, their password. Do we want to support that as a common case or should we fall back to emailing?

### PR DESCRIPTION
**Pull request by Ninjaclasher**
Fixes #1996.

Request email change process:
1. User goes to email change endpoint with a form that consists of a password and email. This endpoint is accessible via edit profile.
2. Verify password and verify email is not used or blacklisted.
3. Create a base64-encoded key that is {user id, new email} signed by SECRET_KEY and create an endpoint that accepts this key. This key will contains a timestamp, which will be used to check expiry.
4. Send this URL to the new email, send a security notice to the old email (email change requested).

Confirm email change process:
1. User clicks the URL in the email.
2. Verify signing, verify expiry timestamp, verify user id matches currently logged in user.
3. Verify again that new email is not used and update email in the database (atomically).

### CODE REVIEW

**Comment by kiritofeng** [comment]
Can we also add a 2FA check before changing the email? (I believe the password change form doesn't require 2FA verification but I think we should also add 2FA to that).

**Comment by Ninjaclasher** [comment]
After some more thought, the issue with requiring 2fa is that we need to support all the different types of 2fa (backup code, totp, hardware key), which requires quite a bit of refactoring. Also, at the moment, adding 2fa won't increase security, since a user can regenerate their totp token/get new backup codes without any re-authentication.

I think adding 2fa to this should be part of a future change where we harden everything together, and is out of the scope of the current PR.

File: judge/views/user.py
Comment (kiritofeng) [inline judge/views/user.py:616]: Should this go outside the revision?

File: judge/forms.py
Comment (kiritofeng) [inline judge/forms.py:109]: What about deactivated accounts?

File: templates/registration/email_change.html
Comment (kiritofeng) [inline templates/registration/email_change.html:1]: Would it be possible to see a screenshot of these emails (mostly to check they render correctly and don't have funny messed up links.)

File: dmoj/settings.py
Comment (kiritofeng) [inline dmoj/settings.py:113]: Might want to add comments for units / explaining purpose (my first thought was `DMOJ_EMAIL_CHANGE_LIMIT_COUNT` was a hard cap on the number of times you could change your email in total.

File: judge/views/user.py
Comment (kiritofeng) [inline judge/views/user.py:530]: Should this method be here, or should we refactor it out into some util file?

**Review by kiritofeng — COMMENTED** [review]
Would also like @quantum5 and/or @Xyene to take a look.

File: judge/forms.py
Comment (Ninjaclasher) [reply judge/forms.py:109]: I believe the current check (during registration) includes all users, including deactivated ones. Should we allow users to reuse emails from deactivated emails?

Comment (kiritofeng) [reply judge/forms.py:109]: I think a better solution might be to set the emails of deactivated accounts to a dummy email and leave the current checks intact, cc @quantum5 @Xyene for thoughts.

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py:616]: The idea was to make this atomic (since the revision block is atomic), so we avoid any races between when we check and when we update the email address. Email uniqueness is not guaranteed on the database level, so we have to do this. If you have any suggestions for how to make it better, please let me know.

**Comment by Ninjaclasher** [comment]
Screenshots
image
image
image
image

File: templates/registration/email_change.html
Comment (Ninjaclasher) [reply templates/registration/email_change.html:1]: I've attached some screenshots of both the HTML email and the plaintext email.

File: judge/views/user.py
Comment (kiritofeng) [reply judge/views/user.py:616]: That is a good point; race conditions hard :(

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py]: Good point, I've refactored it.

File: dmoj/settings.py
Comment (Riolku) [inline dmoj/settings.py:118]: I'm mildly more in favour of a default around 1 hour.

File: judge/forms.py
Comment (Riolku) [inline judge/forms.py:101]: Obviously, all this explodes for social auth. Do we have a notice about this? What if a user has social auth but also a password?

File: judge/utils/mail.py
Comment (Riolku) [inline judge/utils/mail.py:13]: What does the email RFC say about multiple @ in email addresses?

File: judge/forms.py
Comment (Riolku) [inline judge/forms.py:113]: The function name doesnt match the error message. Rename the function to something like `is_email_address_bad_domain` maybe.

File: judge/views/register.py
Comment (Riolku) [inline judge/views/register.py:45]: Let's refactor this common code. Potentially make a function called `validate_not_bad_email_domain` or something.

File: judge/views/user.py
Comment (Riolku) [inline judge/views/user.py:549]: Why is from_email None? Also, maybe we should mark the parameter so it's clear what is being passed as None.

File: judge/views/user.py
Comment (Riolku) [inline judge/views/user.py:590]: This request has expired

File: judge/views/user.py
Comment (Riolku) [inline judge/views/user.py:593]: I feel like an email change confirmation shouldn't require a login. I often click on confirmation links on my phone that i requested on my computer.

File: judge/views/user.py
Comment (Riolku) [inline judge/views/user.py:603]: When does this happen?

**Review by Riolku — COMMENTED** [review]
Will have to review HTML another time, or leave to someone else.

File: templates/registration/email_change_notify_email.html
Comment (Riolku) [inline templates/registration/email_change_notify_email.html:20]: Shouldnt we suggest changing pwd as well?

File: templates/registration/email_change_activate_email.html
Comment (Riolku) [inline templates/registration/email_change_activate_email.html:1]: Is this HTML copied from somewhere?

Comment (Ninjaclasher) [reply templates/registration/email_change_activate_email.html:1]: Yes, it's copied from the password reset/account activation emails.

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py:603]: Any of the above errors that get raised are caught here and returned as a generic message.

File: judge/views/register.py
Comment (Ninjaclasher) [reply judge/views/register.py:45]: This is what I had originally, but I dislike having a function `validate_...` that doesn't return anything and can only raise. I'm open to other suggestions though.

File: judge/utils/mail.py
Comment (Ninjaclasher) [reply judge/utils/mail.py:13]: It looks like you can only have multiple @ in the username part of the email address, while the domain part cannot contain a @. I think this should be fine since we can assume anything after the last part is the domain.

File: judge/views/register.py
Comment (Ninjaclasher) [reply judge/views/register.py:45]: On further thought, I'll change back to `validate_not_bad_email_domain`, since it looks like Django itself uses a similar construct.

File: templates/registration/email_change_notify_email.html
Comment (Ninjaclasher) [reply templates/registration/email_change_notify_email.html]: Good idea, updated.

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py]: Done, also added a comment to explain why. When `from_email` is none, then the `DEFAULT_FROM_EMAIL` setting is used.

File: judge/forms.py
Comment (Ninjaclasher) [reply judge/forms.py:101]: We don't right now, but we also don't have anything set up for password change. In a future PR, we should just hide these endpoints if the user has no password set and say something about using social auth.

Right now, I'm of the opinion that if a user is using social auth but has a password set, then they should be treated as a normal user and be allowed to change their emails.

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py:594]: Hmm, I'm open to discussion on this. I do think there's some merit to ensuring the user is logged in before switching emails. This prevents a user mistyping an email that is actually in use, and then the user of the mistyped email accidentally clicking on the link causing the email to be switched.

File: judge/forms.py
Comment (Riolku) [reply judge/forms.py:101]: What happens when a user with social auth changes their email? Does their social auth still work? If so I think it's probably fine.

File: judge/views/register.py
Comment (Riolku) [reply judge/views/register.py]: I think `throw_if_bad_email_domain` could work too, but don't feel like bikeshedding much more.

File: judge/forms.py
Comment (Ninjaclasher) [reply judge/forms.py:101]: Yes it should be fine. I don't believe social auth uses this email field after initially populating it.

File: judge/views/user.py
Comment (Riolku) [reply judge/views/user.py:594]: Like you mean user A requests a change to email B, and the owner of B can now take over the account of A...

Ok, granted, that seems bad. I think this is fine then.

File: judge/views/user.py
Comment (Riolku) [reply judge/views/user.py:604]: Do we have this pattern elsewhere?

I'm tempted to say we should move the logic of changing to its own function, and make the displaying be in `get()`.

Also, is there a method that makes more sense than `str(ve)`

Comment (Riolku) [reply judge/views/user.py:604]: Finally can we use a dedicated exception instead of `ValueError`? I thought we failed some conversion or smth.

Comment (Ninjaclasher) [reply judge/views/user.py]: Changed to a `EmailChangedError`. I think `str(e)` is the standard way to retrieve the message from an error, unless I am mistaken.

Comment (Riolku) [reply judge/views/user.py]: Are you against moving the logic into its own function to avoid the double try?

Comment (Ninjaclasher) [reply judge/views/user.py]: I don't really see the point, since it'll only ever be used once, but I'll move it to a separate function if it makes it easier to read the code.

Comment (Riolku) [reply judge/views/user.py]: Other people can comment on this.. I could live with the current code.

**Review by Riolku — APPROVED** [review]
LGTM aside from the minor bikeshedding on code readability.

EDIT: Oops seems I missed the recent update.

Either way, thanks for working on this!

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py]: I've moved it to a separate function.

File: judge/forms.py
Comment (Xyene) [reply judge/forms.py:109]: Agree, I think this mirroring the registration logic is good.

File: judge/utils/mail.py
Comment (Xyene) [inline judge/utils/mail.py:1]: Since this is brand new code, let's add type annotations.

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:550]: If you are going to pass these as kwargs, then you should make it explicitly required to do so by adding `*` to the parameter list of `send_mail`.

File: judge/utils/mail.py
Comment (Xyene) [inline judge/utils/mail.py:23]: The ordering of these parameters is a little strange, it'd be good to keep everything template-related together, not separated by email parameters.

File: dmoj/settings.py
Comment (Xyene) [inline dmoj/settings.py:115]: Probably makes sense for these fields to all be in minutes.

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:595]:
```suggestion
                _('Please try again while logged in to the account this email change was originally requested from.'),
```

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:605]: We could save people a click by embedding the to/from emails here.

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:600]:
```suggestion
                    _('The email you originally requested has since been registered by another user. Please try again with a new email.'),
```

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:610]: Is the `self.` here actually necessary?

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:615]:
```suggestion
                _('Email successfully changed'),
```

File: judge/views/user.py
Comment (Xyene) [inline judge/views/user.py:616]: If it is easy (and I think we have the data here?) we should say to what in this message.

File: templates/registration/email_change_activate_email.html
Comment (Xyene) [inline templates/registration/email_change_activate_email.html:12]: Let's just assert that the expiry is > 1 second as part of config validation, and avoid needing to pluralize this. 1 minute expiry is unusable UX anyway.

File: templates/registration/email_change_activate_email.html
Comment (Xyene) [inline templates/registration/email_change_activate_email.html:9]:
```suggestion
{% trans %}Someone (hopefully you!) has requested to change the email address for their user account at {{ site_name }} to this email.{% endtrans %}
```

File: templates/registration/email_change_notify_email.html
Comment (Xyene) [inline templates/registration/email_change_notify_email.html:9]:
```suggestion
{% trans %}Someone (hopefully you!) has requested to change the email address associated with your user account at {{ site_name }} to {{ new_email }}. {% endtrans %}
```

File: templates/registration/email_change_notify_email.html
Comment (Xyene) [inline templates/registration/email_change_notify_email.html:24]:
```suggestion
```

File: templates/registration/email_change_notify_email.txt
Comment (Xyene) [inline templates/registration/email_change_notify_email.txt:2]:
```suggestion
{% trans %}Someone (hopefully you!) has requested to change the email address associated with your user account at {{ site_name }} to {{ new_email }}. {% endtrans %}
```

File: judge/views/user.py
Comment (Ninjaclasher) [reply judge/views/user.py:614]: Yes, otherwise we get a `NameError` (flake8 also complains)

File: judge/utils/mail.py
Comment (Ninjaclasher) [reply judge/utils/mail.py:23]: This was originally copied from https://github.com/django/django/blob/main/django/contrib/auth/forms.py#L287. I think their parameters were ordered that way since the HTML template was optional. Since we want likely want our emails to always contain an HTML version, I will reorder the parameters.

File: templates/registration/email_change_activate_email.html
Comment (Ninjaclasher) [reply templates/registration/email_change_activate_email.html:12]: I've added an assertion for > 1 minute expiry, but going to wait on @quantum5's reply before changing the template to verify this will work for other languages.

Comment (Ninjaclasher) [reply templates/registration/email_change_activate_email.html:12]: Seems like this is necessary for other languages, according to @int-y1.

**Review by Xyene — APPROVED** [review]
LGTM, could you please rebase and squash?

@quantum5, could you also take a quick look over this before I merge it?

File: templates/registration/email_change.html
Comment (quantum5) [inline templates/registration/email_change.html:10]: Maybe move this into a CSS file eventually, but who cares right now.

File: templates/registration/email_change_activate_email.html
Comment (quantum5) [reply templates/registration/email_change_activate_email.html:12]: Yes, lack of ngettext is going to be so much fun in a bunch of languages...

**Review by quantum5 — APPROVED** [review]
LGTM, get this merged soon so we don't have to deal with all those manual requests.

**Pull request merged**
